lib/spack/spack/environment/environment.py

```diff
@@ -1989,11 +1978,14 @@ def get_one_by_hash(self, dag_hash):
 
     def all_matching_specs(self, *specs: spack.spec.Spec) -> List[Spec]:
         """Returns all concretized specs in the environment satisfying any of the input specs"""
-        key = lambda s: s.dag_hash()
+        # Look up abstract hashes ahead of time, to avoid O(n^2) traversal.
+        specs = [s.lookup_hash() for s in specs]
+
+        # Avoid double lookup by directly calling _satisfies.
```

Comment:
This is fine in this PR, but we need to re-think the interface.

Reply:
Yeah, the nice solution is to make lookup context dependent instead of making it rely on globals. I've advocated for that in the abstract hash PR too. In this case, we wanna find stuff in an environment, so it makes no sense to "globally" resolve the hash, where it may even cause a binary cache index to be downloaded...

If I do `spack -e . install /xyz` it should look up `/xyz` in the environment and nowhere else, therefore I think it's wrong to "resolve" the hash whenever you hit satisfies...

Reply:
And also: when you have `specs: [/xyz]` of course it doesn't make sense to look up this hash in the environment, but it should be looked up in the database / externally. So again, this global lookup logic doesn't make any sense, and this change ^ is just a hotfix.